crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1401,6 +1406,112 @@ impl<'db> Type<'db> {
+
+            (Type::ModuleLiteral(first), Type::ModuleLiteral(second)) => {
+                first.module(db) == second.module(db)
+            }
+
+            (Type::IntLiteral(first), Type::IntLiteral(second)) => first == second,
+            (Type::BooleanLiteral(first), Type::BooleanLiteral(second)) => first == second,
+            (Type::StringLiteral(first), Type::StringLiteral(second)) => {
+                first.value(db).as_ref() == second.value(db).as_ref()
+            }
+            (Type::BytesLiteral(first), Type::BytesLiteral(second)) => {
+                first.value(db).as_ref() == second.value(db).as_ref()
+            }
+            (Type::SliceLiteral(first), Type::SliceLiteral(second)) => {
+                first.as_tuple(db) == second.as_tuple(db)
+            }
+
+            (Type::ClassLiteral(first), Type::ClassLiteral(second)) => first.class == second.class,
+            (Type::Instance(first), Type::Instance(second)) => first.class == second.class,
+
+            (Type::SubclassOf(first), Type::SubclassOf(second)) => {
+                match (first.base, second.base) {
+                    (ClassBase::Todo(_), ClassBase::Todo(_)) => false,
+                    (ClassBase::Any | ClassBase::Unknown, ClassBase::Any | ClassBase::Unknown) => {
+                        true
+                    }
+                    (ClassBase::Class(first), ClassBase::Class(second)) => first == second,
+                    _ => false,
+                }
+            }
+
+            (Type::Tuple(first), Type::Tuple(second)) => {
+                first.len(db) == second.len(db)
+                    && iter::zip(first.elements(db), second.elements(db)).all(equal)
+            }
+
+            (Type::Union(first), Type::Union(second)) => {
+                let first_elements = first.elements(db);
+                let second_elements = second.elements(db);
+
+                if first_elements.len() != second_elements.len() {
+                    return false;
+                }
+
+                let first_elements = first_elements.iter().collect::<OrderedTypeSet>();
+                let second_elements = second_elements.iter().collect::<OrderedTypeSet>();
+
+                iter::zip(first_elements, second_elements).all(equal)
+            }
+
+            (Type::Intersection(first), Type::Intersection(second)) => {
+                let first_positive = first.positive(db).iter().collect::<OrderedTypeSet>();
+                let first_negative = first.negative(db).iter().collect::<OrderedTypeSet>();
+
+                let second_positive = second.positive(db).iter().collect::<OrderedTypeSet>();
+                let second_negative = second.negative(db).iter().collect::<OrderedTypeSet>();
+
+                iter::zip(first_positive, second_positive).all(equal)
+                    && iter::zip(first_negative, second_negative).all(equal)
+            }
```

Comment:
Nice; this is a fix we need in `is_equivalent_to` as well. It would be nice if there were a way to do it without allocating so many hashsets...